### ISSUE TITLE
Neo.DatabaseError.General.UnknownError: key not found: Variable(k)

### ISSUE THREAD

**Issue by CharlieDigital** [neo4j/neo4j#12126]
Error occurs in 3.5.1 with a blank database and easily reproducible with the following nonsense query:

```
MATCH (a:A)--(b:B)
OPTIONAL MATCH (b)--(c:C)--(d:D)
WITH a, b, c, a AS k
WHERE c IS NOT NULL OR (d IS NULL AND k <> b)
RETURN a
```

Even simpler version:

```
MATCH (a:A)
WITH a, a AS k
WHERE a.X = 1 OR k <> a
RETURN a
```

- Neo4j version: 3.5.1
- Operating system: Windows Server 2016
- API/Driver: None; browser client
- **Steps to reproduce**
  - In empty 3.5.1 database, paste the query above and execute
  - Observe the error
  - Remove the first `OR` condition
  - Observe there is no error
- Expected behavior: query produces no results.
- Actual behavior: query engine throws error; variable "k" is not provided to the execution context of the conditional evaluator in the `WHERE` clause despite being declared above.  Seems to be an issue specific to negation and node comparisons.

Additional Notes: when the first part of the `OR` condition is removed (delete `c IS NOT NULL OR`), query returns empty resultset as expected.

Only seems to occur with the `<>` operator; when using `=` in the second clause, comparison occurs as expected with no error.

Using `NOT (k = a)` also triggers the error.

Workaround is to compare some property of `k` and `a` instead of performing the node comparison.  This results in more DB hits and lower performance, but can be used to work around the issue.  For example: `a.Z <> k.Z`.  The workaround for #9347 also works: prefix query with `CYPHER 3.1`.

Similar to #9347, but does not require `DISTINCT` keyword

**Comment by Hunterness** [comment]
Hi, thanks for reporting, we will look into it.

**Comment by Lojjs** [comment]
@CharlieDigital We have got another report about this issue and thus already fixed it. The fix will be included in the 3.5.2 release which will most likely be available sometime during this week.
